pkgs/build-support/build-fhs-userenv/chrootenv.c

```diff
@@ -0,0 +1,182 @@
+                         "DBUS_SESSION_BUS_ADDRESS"};
+
+char **env_build(char *names[], size_t len) {
+  char *env, **ret = malloc((len + 1) * sizeof(char *)), **ptr = ret;
+
+  for (size_t i = 0; i < len; i++) {
+    if ((env = getenv(names[i]))) {
+      if (asprintf(ptr++, "%s=%s", names[i], env) < 0)
+        errorf(EX_OSERR, "asprintf");
+    }
+  }
+
+  *ptr = NULL;
+  return ret;
+}
+
+struct bind {
+  char *from;
+  char *to;
+};
+
+struct bind binds[] = {{"/", "host"},   {"/proc", "proc"}, {"/sys", "sys"},
+                       {"/nix", "nix"}, {"/tmp", "tmp"},   {"/var", "var"},
+                       {"/run", "run"}, {"/dev", "dev"},   {"/home", "home"}};
+
+void bind(struct bind *bind) {
+  DIR *src = opendir(bind->from);
+
+  if (src) {
+    if (closedir(src) < 0)
+      errorf(EX_IOERR, "closedir");
+
+    if (mkdir(bind->to, 0755) < 0)
+      errorf(EX_IOERR, "mkdir");
+
+    if (mount(bind->from, bind->to, "bind", MS_BIND | MS_REC, NULL) < 0)
+      errorf(EX_OSERR, "mount");
+
+  } else {
+    // https://github.com/NixOS/nixpkgs/issues/31104
+    if (errno != ENOENT)
+      errorf(EX_OSERR, "opendir");
+  }
+}
+
+void spitf(char *path, char *fmt, ...) {
+  va_list args;
+  va_start(args, fmt);
+
+  FILE *f = fopen(path, "w");
+
+  if (f == NULL)
+    errorf(EX_IOERR, "spitf(%s): fopen", path);
+
+  if (vfprintf(f, fmt, args) < 0)
+    errorf(EX_IOERR, "spitf(%s): vfprintf", path);
+
+  if (fclose(f) < 0)
+    errorf(EX_IOERR, "spitf(%s): fclose", path);
+}
+
+int nftw_rm(const char *path, const struct stat *sb, int type,
+            struct FTW *ftw) {
+  if (remove(path) < 0)
+    errorf(EX_IOERR, "nftw_rm");
+
+  return 0;
+}
+
+#define LEN(x) sizeof(x) / sizeof(*x)
+
+int main(int argc, char *argv[]) {
+  if (argc < 2) {
+    fprintf(stderr, "Usage: %s command [arguments...]\n"
+                    "Requires Linux kernel >= 3.19 with CONFIG_USER_NS.\n",
+            argv[0]);
+    exit(EX_USAGE);
+  }
+
+  char tmpl[] = "/tmp/chrootenvXXXXXX";
+  char *root = mkdtemp(tmpl);
+
+  if (root == NULL)
+    errorf(EX_IOERR, "mkdtemp");
+
+  // Don't make root private so that privilege drops inside chroot are possible:
+  if (chmod(root, 0755) < 0)
+    errorf(EX_IOERR, "chmod");
+
+  pid_t cpid = fork();
+
+  if (cpid < 0)
+    errorf(EX_OSERR, "fork");
+
+  if (cpid == 0) {
+    uid_t uid = getuid();
+    gid_t gid = getgid();
+
+    // If we are root, no need to create new user namespace.
+    if (uid == 0) {
+      if (unshare(CLONE_NEWNS) < 0)
+        errorf(EX_OSERR, "unshare");
+      // Mark all mounted filesystems as slave so changes
+      // don't propagate to the parent mount namespace.
+      if (mount(NULL, "/", NULL, MS_REC | MS_SLAVE, NULL) < 0)
+        errorf(EX_OSERR, "mount");
+    } else {
+      // Create new mount and user namespaces. CLONE_NEWUSER
+      // requires a program to be non-threaded.
+      if (unshare(CLONE_NEWNS | CLONE_NEWUSER) < 0)
+        errorf(EX_OSERR, "unshare");
```

Comment:
`unshare` is not allowed on all linux distributions by default. On redhat and clones, it is required to set `kernel.unprivileged_userns_clone = 1`. Maybe the error message can reflect that if the return value of `unshare` is EPERM.

Reply:
Thank you, good catch. Also Arch Linux has it disabled by default in kernel (requires `linux-userns` kernel from AUR), but Debian variants should be OK. Do you think it's appropriate to check for distro-specific files (like `/etc/redhat-release`) and print specific instructions?

Reply:
I would keep the error message generic to keep the code simple:

```
No permission to create username spaces as unprivileged user.
Please make sure that the kernel has usernamespace support enabled (`CONFIG_USER_NS=y`)
and it is enabled for unprivileged users (sysctl -w kernel.unprivileged_userns_clone=1).
```

Reply:
The problem is that I think `kernel.unprivileged_userns_clone` isn't in Torvalds' tree and most likely a RHEL-specific patch on top of it:

```
$ find /proc/sys/kernel -name unprivileged_\*
/proc/sys/kernel/unprivileged_bpf_disabled
/proc/sys/kernel/unprivileged_userns_apparmor_policy
```

Reply:
It is also available on ubuntu.

Reply:
This would also fail on old Linux kernels. We may want to have a nice generic error message, in lines of "You may have an old kernel or have CLONE_NEWUSER disabled by your distribution security settings."

Reply:
@abbradar, @Mic92: Got you. Hope they enable it by default at some point (and old kernels will be phased out of use), but for now going to print this stuff, supplemented by `sysctl` instructions if `/proc/sys/kernel/unprivileged_userns_clone` is present.

Arch Linux should have adopted this patch instead of disabling the option entirely...